engine/crates/integration-tests/tests/federation/auth/authenticated.rs

```diff
@@ -0,0 +1,97 @@
+use futures::Future;
+use gateway_v2::Gateway;
+use graphql_mocks::{MockGraphQlServer, SecureSchema};
+use integration_tests::{
+    federation::{GatewayV2Ext, TestFederationGateway},
+    openid::{CoreClientExt, OryHydraOpenIDProvider, JWKS_URI},
+    runtime,
+};
+
+pub(super) fn with_secure_schema<F, O>(f: impl FnOnce(TestFederationGateway) -> F) -> O
+where
+    F: Future<Output = O>,
+{
+    runtime().block_on(async move {
+        let secure_mock = MockGraphQlServer::new(SecureSchema::default()).await;
+
+        let engine = Gateway::builder()
+            .with_schema("secure", &secure_mock)
+            .await
+            .with_supergraph_config(format!(
+                r#"extend schema @authz(providers: [
+                {{ name: "my-jwt", type: jwt, jwks: {{ url: "{JWKS_URI}" }} }},
+                {{ type: anonymous }}
+            ])"#
+            ))
+            .finish()
+            .await;
+
+        f(engine).await
+    })
+}
+
+#[test]
+fn authenticated() {
+    with_secure_schema(|engine| async move {
+        let token = OryHydraOpenIDProvider::default()
+            .create_client()
+            .await
+            .get_access_token_with_client_credentials(&[])
+            .await;
+
+        let response = engine
+            .execute("query { anonymous }")
+            .header("Authorization", format!("Bearer {token}"))
+            .await;
+        insta::assert_json_snapshot!(response, @r###"
+        {
+          "data": {
+            "anonymous": "Hello anonymous!"
+          }
+        }
+        "###);
+
+        let response = engine
+            .execute("query { mustBeAuthenticated }")
+            .header("Authorization", format!("Bearer {token}"))
+            .await;
+        insta::assert_json_snapshot!(response, @r###"
+        {
+          "data": {
+            "mustBeAuthenticated": "You are authenticated"
+          }
+        }
+        "###);
+    });
+}
+
+#[test]
+fn not_authenticated() {
```

Comment:
I'm reading the tests first so maybe it becomes later, but if only part of the query is on `@authenticated` fields and the rest is not, are we still returning data with nulls for the authenticated fields? This test is correct because the whole query would be null'ed.

Reply:
hum, excellent point... Excellent catch. Definitely should be.